libdnf/module/module_metadata.cpp

```diff
@@ -0,0 +1,168 @@
+            g_object_ref(module_merger);
+        }
+    }
+    return *this;
+}
+
+
+ModuleMetadata::~ModuleMetadata() {
+    if (resulting_module_index != nullptr) {
+        g_object_unref(resulting_module_index);
+    }
+    if (module_merger != nullptr) {
+        g_object_unref(module_merger);
+    }
+}
+
+
+void ModuleMetadata::add_metadata_from_string(const std::string & yaml, int priority) {
+    GError * error = NULL;
+    g_autoptr(GPtrArray) failures = NULL;
+
+    ModulemdModuleIndex * module_index = modulemd_module_index_new();
+    gboolean success = modulemd_module_index_update_from_string(module_index, yaml.c_str(), FALSE, &failures, &error);
+    if (!success) {
+        auto & logger = *base->get_logger();
+        for (unsigned int i = 0; i < failures->len; i++) {
+            ModulemdSubdocumentInfo * item = (ModulemdSubdocumentInfo *)(g_ptr_array_index(failures, i));
+            // TODO(pkratoch): Maybe use LogEvent instead
+            logger.error("Module yaml error: {}", modulemd_subdocument_info_get_gerror(item)->message);
+        }
+    }
+    if (error) {
+        throw ModuleResolveError(M_("Failed to update from string: {}"), error->message);
+    }
+    if (!module_merger) {
+        module_merger = modulemd_module_index_merger_new();
+        if (resulting_module_index) {
+            // Priority is set to 0 in order to use the current resulting_module_index data as a baseline
+            modulemd_module_index_merger_associate_index(module_merger, resulting_module_index, 0);
+            g_clear_pointer(&resulting_module_index, g_object_unref);
+        }
+    }
+
+    modulemd_module_index_merger_associate_index(module_merger, module_index, priority);
+    g_object_unref(module_index);
+}
+
+
+void ModuleMetadata::resolve_added_metadata() {
+    if (!module_merger) {
+        return;
+    }
+
+    GError * error = NULL;
+
+    resulting_module_index = modulemd_module_index_merger_resolve(module_merger, &error);
+    if (error && !resulting_module_index) {
+        throw ModuleResolveError(M_("Failed to resolve: {}"), (error->message) ? error->message : "Unknown error");
+    }
+    if (error) {
+        auto & logger = *base->get_logger();
+        // TODO(pkratoch): Maybe use LogEvent instead
+        logger.debug("There were errors while resolving modular defaults: {}", error->message);
+    }
+
+    modulemd_module_index_upgrade_defaults(resulting_module_index, MD_DEFAULTS_VERSION_ONE, &error);
+    if (error) {
+        throw ModuleResolveError(M_("Failed to upgrade defaults: {}"), error->message);
+    }
+    modulemd_module_index_upgrade_streams(resulting_module_index, MD_MODULESTREAM_VERSION_TWO, &error);
+    if (error) {
+        throw ModuleResolveError(M_("Failed to upgrade streams: {}"), error->message);
+    }
+    g_clear_pointer(&module_merger, g_object_unref);
+}
+
+
+std::pair<std::vector<ModuleItem *>, std::vector<ModuleItem *>> ModuleMetadata::get_all_module_items() {
```

Comment:
I am not sure about the implementation of the method, may be we can discuss it directly.